drivers/usrsock/usrsock_rpmsg_server.c

```diff
@@ -1042,28 +1068,84 @@ static void usrsock_rpmsg_ns_unbind(FAR struct rpmsg_endpoint *ept)
   kmm_free(ept);
 }
 
-static int usrsock_rpmsg_ept_cb(FAR struct rpmsg_endpoint *ept,
-                                FAR void *data, size_t len, uint32_t src,
-                                FAR void *priv_)
+static int usrsock_rpmsg_ept_do_cb(FAR struct usrsock_rpmsg_ept_s *uept,
+                                   FAR void *data, size_t len, uint32_t src,
+                                   FAR struct usrsock_rpmsg_s *priv)
 {
   FAR struct usrsock_request_common_s *common = data;
-  FAR struct usrsock_rpmsg_ept_s *uept =
-    (FAR struct usrsock_rpmsg_ept_s *)ept;
-  FAR struct usrsock_rpmsg_s *priv = priv_;
 
   if (uept->remain > 0)
     {
-      return usrsock_rpmsg_sendto_handler(ept, data, len, src, priv);
+      return usrsock_rpmsg_sendto_handler(&uept->ept, data, len, src, priv);
     }
   else if (common->reqid >= 0 && common->reqid <= USRSOCK_REQUEST__MAX)
     {
-      return g_usrsock_rpmsg_handler[common->reqid](ept, data, len,
+      return g_usrsock_rpmsg_handler[common->reqid](&uept->ept, data, len,
                                                     src, priv);
     }
 
   return -EINVAL;
 }
 
+static int usrsock_rpmsg_ept_cb(FAR struct rpmsg_endpoint *ept,
+                                FAR void *data, size_t len, uint32_t src,
+                                FAR void *priv_)
+{
+  FAR struct usrsock_rpmsg_s *priv = priv_;
+  FAR struct usrsock_rpmsg_ept_s *uept =
+    (FAR struct usrsock_rpmsg_ept_s *)ept;
+  FAR struct usrsock_rpmsg_req_s *req;
+  int ret;
+
+  /* This callback is called from only one thread per ept, so don't need any
+   * lock for `inuse` state.
+   */
+
+  if (uept->inuse)
+    {
+      /* Avoid recursive call. */
+
+      req = (FAR struct usrsock_rpmsg_req_s *)sq_remfirst(&uept->req_free);
+      if (req == NULL)
+        {
+          return -ENOMEM;
+        }
+
+      req->data = data;
+      req->len  = len;
+      req->src  = src;
+      sq_addlast(&req->flink, &uept->req_pending);
```

Comment:
should we use double link to avoid the search?

Reply:
`sq_addlast` doesn't need any search, it just adds the node after the `tail` in `sq_queue_t`.

Reply:
Ok, you are right.